### PR TITLE
[bitnami/thanos] feat: create sharded hpa and pdb for storegateway

### LINKED ISSUE #11305: [thanos] Store Gateway HPA does not work with sharding enabled
### Name and Version

bitnami/thanos 11.1.0

### What steps will reproduce the bug?

Install the chart with the given values.

### Are you using any custom parameters or values?
```yaml
storegateway:
  autoscaling:
    enabled: true
  sharded
    enabled: true
```

### What is the expected behavior?

There should be one HPA per shard. 

### What do you see instead?

```shell
$ kubectl get events
115s        Warning   FailedGetScale                 horizontalpodautoscaler/thanos-storegateway                     statefulsets.apps "thanos-storegateway" not found

$ kubectl get hpa
thanos-storegateway     StatefulSet/thanos-storegateway    <unknown>/80%   3         6         0          3h22m

$ kubectl get sts
thanos-storegateway-0                             1/1     3h22m
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by aslafy-z**
### Description of the change

Fixes an issue with the storegateway HPA/PDB not being well generated in case of a sharded configuration.

### Benefits

The functionality will be fixed.

### Possible drawbacks

People that missed this issue and upgrade to the latest version with sharding enabled will see storegateway start scaling. 

### Applicable issues

  - fixes #11305

### Checklist

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/). This is *not necessary* when the changes only affect README.md files.
- [x] Title of the pull request follows this pattern [bitnami/<name_of_the_chart>] Descriptive title
- [x] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/master/CONTRIBUTING.md#sign-your-work)

### CODE REVIEW

**Comment by rafariossaa** [comment]
Hi, it is giving an error in the linter:
````
==> Linting /tmp/puzzle-workplace-6637326479459654374/extracted/aslafy-z-charts-1-4d65686/bitnami/thanos
[ERROR] templates/: parse error at (thanos/templates/storegateway/hpa-sharded.yaml:59): unexpected {{end}}

Error: 1 chart(s) linted, 1 chart(s) failed
```
Could you take a look to it ?

You can find the report in the [artifacts generated](https://github.com/bitnami/charts/actions/runs/2801918811#artifacts):

File: bitnami/thanos/templates/storegateway/hpa-sharded.yaml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        name: cpu
+        {{- if semverCompare "<1.23-0" (include "common.capabilities.kubeVersion" .) }}
+        targetAverageUtilization: {{ .Values.storegateway.autoscaling.targetCPU }}
+        {{- else }}
+        target:
+          type: Utilization
+          averageUtilization: {{ .Values.storegateway.autoscaling.targetCPU }}
+        {{- end }}
+    {{- end }}
+{{- end }}
```
Comment (aslafy-z) [inline bitnami/thanos/templates/storegateway/hpa-sharded.yaml:56]:
```suggestion
```

**Comment by aslafy-z** [comment]
@rafariossaa it should now be fixed. Thanks for the heads up

File: bitnami/thanos/templates/storegateway/hpa-sharded.yaml [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+{{- else }}
+  {{- $shards = len .Values.storegateway.sharded.timePartitioning }}
+{{- end }}
+
+{{- range $index, $_ := until $shards }}
+apiVersion: {{ include "common.capabilities.hpa.apiVersion" ( dict "context" $ ) }}
+kind: HorizontalPodAutoscaler
+metadata:
+  name: {{ printf "%s-storegateway-%s" (include "common.names.fullname" $) (toString $index) | trunc 63 | trimSuffix "-" }}
+  namespace: {{ .Release.Namespace | quote }}
```
Comment (joancafom) [inline bitnami/thanos/templates/storegateway/hpa-sharded.yaml:15]: As you are looping over a `range` function, the scope is changed and `.Release` & `.Values` references can no longer be resolved. You should use `$.Release` and `$.Values` instead.

See the following examples:

https://github.com/bitnami/charts/blob/2a88f4661df35140dc39c72e7677357de764bcc3/bitnami/thanos/templates/storegateway/service-sharded.yaml#L16

https://github.com/bitnami/charts/blob/2a88f4661df35140dc39c72e7677357de764bcc3/bitnami/thanos/templates/storegateway/service-sharded.yaml#L19

**Review by joancafom — CHANGES_REQUESTED** [review]
Thanks for your PR! I left you a comment, could you please review it?

File: bitnami/thanos/templates/storegateway/hpa-sharded.yaml (hunk 2 above)
Comment (aslafy-z) [reply bitnami/thanos/templates/storegateway/hpa-sharded.yaml]: I just updated the resource and tested the result was the expected one.

**Comment by aslafy-z** [comment]
I think there should also be a sharded version of the storegateway PDB. Would you let me adding it to this PR? Also, what label should I add for the PDB to discover the pod resources? `shard=1` would do the job? I'd prefer not to use the `component` label to store the shard number. Do you see any other *clean* option?

**Comment by joancafom** [comment]
Hi @aslafy-z, sorry for the delay

>I think there should also be a sharded version of the storegateway PDB. Would you let me adding it to this PR? 

Yes, totally! It would be nice if you added it to the PR 😁 . After that, you would only need to update the PR description accordingly.

>What label should I add for the PDB to discover the pod resources? maybe shard=N could do the job? I'd prefer not to use the component label to store the shard number. WDYT?

You can use a composition of `component` & `shard`:

```diff
  selector:
    matchLabels: {{- include "common.labels.matchLabels" . | nindent 6 }}
      app.kubernetes.io/component: storegateway
+     shard: {{ $i | quote }}
```

This works nicely :) 
```
$ kubectl get pod -l "app.kubernetes.io/component=storegateway,shard=0"
NAME                      READY   STATUS    RESTARTS   AGE
thanos-storegateway-0-0   1/1     Running   0          64m
```

**Comment by aslafy-z** [comment]
@joancafom Should be good by now, I also added shard labels to other objects than pods.

**Review by joancafom — CHANGES_REQUESTED** [review]
Great, thanks for your changes again. Just one more tiny detail and I think we are good to go!

File: bitnami/thanos/Chart.yaml [hunk 3]
```diff
@@ -28,4 +28,4 @@ name: thanos
 sources:
   - https://github.com/bitnami/containers/tree/main/bitnami/thanos
   - https://thanos.io
-version: 11.2.0
+version: 11.2.1
```
[the remark this answers is not in the document]
Comment (aslafy-z) [reply bitnami/thanos/Chart.yaml]: Done ;)

**Review by joancafom — APPROVED** [review]
LGTM! Thanks for your changes 🚀

**Pull request merged**